### ISSUE TITLE
[ML] Add LibTorch to the CI build images

### ISSUE THREAD

**Issue by davidkyle** [elastic/ml-cpp#1679]
The binary distributions of LibTorch are not built with the optimal flags for ML and have dependencies that are not required for ML use cases. For example distributed training is not required. For this reason developers and build servers should use a tuned manually built version of the libraries. 

The recommended way to build PyTorch is to use Anaconda. Unfortunately Anaconda is not currently available on Apple M1 machines or Linux aarch64 so the build will use CMake directly. The build is also dependent on Python and a number of Python packages.

- [x] **Linux x86_64**
    - [x] Build LibTorch and supporting libraries in Docker build Image and update CI with the new image. Update linux.md setup instructions #1678 #1688 #1696 #1712
- [x] **Linux aarch64 Cross compiled**
    - [x] Build the native binaries on aarch64 and add to the downloaded dependencies. 
    - [x] Publish docker image with the new download dependencies
- [x] **Linux aarch64 native**
    - [x]  Build LibTorch and supporting libraries in Docker build Image #1714
    - [x] Publish docker image #1732
- [x] **Windows**
    - [x]  Build LibTorch separately and make the libraries available to download by the CI workers #1677 #1692
- [x] **macOS x86_64 Native**
    - [x]  Update the instructions in macos.md #1712
- [x] **macOS x86_64 Cross compiled**
    - [x] Build libraries natively on mac hardware then add S3 bucket for downloaded by the CI workers #1685 
    - [x] Build and publish new docker image with libraries downloaded from S3 #1738 

**macOS M1**: TBC

**Comment by droberts195** [comment]
> **macOS M1**: TBC

Support was added in pytorch/pytorch#48275.  That is merged but currently only in the master branch.  Presumably it will go into version 1.8 when that is released.

The gaps between versions 1.4, 1.5, 1.6 and 1.7 were all approximately 3 months.  If 1.8 followed 3 months after 1.7 then it would be released at the end of this month.  Even if it's a little later, hopefully we won't have to wait too much longer before we can upgrade to version 1.8 and create a macOS M1 build.

**Comment by davidkyle** [comment]
**Large Linux Library Size**
libtorch_cpu.so is ~1.4GB when built with the current options. Investigate options to reduce the size:
- Check it is not a debug build
- Can symbols be stripped?
- Are there any other PyTorch build options that will reduce the footprint?

**Comment by droberts195** [comment]
> * Can symbols be stripped?

When we build a release `dev-tools/strip_binaries.sh` should be stripping all the libraries.  It may be best not to strip the version that's installed in `/usr/local/gcc93/lib`, but stick with the policy of stripping when we construct our distribution, because that will make stack traces nicer for developers working on Linux.

Assuming `strip` cuts the size down from 1.4GB to a more sensible size when we construct our distribution we don't necessarily have to worry about the library itself.

Another issue is that the Docker image size for the build image has gone up from 1.75GB to 23.9GB:

```
docker.elastic.co/ml-dev/ml-linux-build                      16                             aeb3b9c975dc        3 days ago          23.9 GB
docker.elastic.co/ml-dev/ml-linux-build                      15                             6a8be82012bf        9 months ago        1.75 GB
```

Obviously some of this is accounted for by the libraries installed in `/usr/local/gcc93/lib`, like `libtorch_cpu.so`, so the size is bound to go up to something over 3.15GB.  But some of it is also accounted for by the other stuff we're leaving on the build image.  Possibly we should uninstall Python from the build image after building PyTorch.  We don't need Python for pure build images.

There's a benefit in removing unnecessary tools from the build Docker image because it has to be downloaded by every CI build on ephemeral hosts and every release manager build.

**Comment by davidkyle** [comment]
Much of the difference in size between builds 15 and 16 is because the `rm -rf /usr/src/*` command failed in the docker build file. There is an extra 14GB in `/usr/src/` that should be removed.

Python and it's libraries are installed in the image which can be removed once PyTorch is built along with CMake for a smaller gain

**Comment by davidkyle** [comment]
**BLAS and OpenMP Dependencies**
The documented PyTorch build process requires installing the `pip` modules `mkl` and `mkl-include` which install libmkl on the system. When invoking `setup.py` CMake discovers the pip install and chooses MKL for the BLAS library and links to it adding a new dependency. This was not known at the time and causes problems for distribution.

#1696 resolves the dependency for Linux but it will have to be fixed for all platforms. E.g on my mac build where I used Anaconda to install the python dependencies libtorch_cpu.dylib has the rpath entry:
```
          cmd LC_RPATH
      cmdsize 48
         path /Users/davidkyle/opt/anaconda3/lib (offset 12)
```

And links to (among others):
```
	@rpath/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
	@rpath/libmkl_intel_lp64.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmkl_intel_thread.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmkl_core.dylib (compatibility version 0.0.0, current version 0.0.0)
```
libomp (OpenMP) should not be required as the distributed functions are not built, hopefully this can be removed by setting the right switches

**Comment by droberts195** [comment]
> And links to (among others)

On the reference build server that was set up using precisely the instructions in `build-setup/macos.md` this doesn't happen:

```
vm-macosx-1014-64-1:lib dave$ otool -L libtorch_cpu.dylib 
libtorch_cpu.dylib:
        @rpath/libtorch_cpu.dylib (compatibility version 0.0.0, current version 0.0.0)
        @rpath/libc10.dylib (compatibility version 0.0.0, current version 0.0.0)
        /System/Library/Frameworks/Accelerate.framework/Versions/A/Accelerate (compatibility version 1.0.0, current version 4.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 800.7.0)
```

It doesn't massively matter what people have on their own laptops, as long as the reference build machines are set up in a way that doesn't distribute unnecessary stuff.

Still, it's probably best to incorporate the extra `MKL` and `BLAS` flags into the macOS build instructions just to ensure no accidents in the future.  They won't make any difference to the reference build server as currently configured, but they might in the future.

**Comment by davidkyle** [comment]
CMake has picked [Acclerate](https://developer.apple.com/documentation/accelerate) for the BLAS library in that case. I will look into making that option explicit so the build is repeatable on developers machines where they are using Anaconda and update the setup instructions to remove `pip install mlk...mkl-include`

**Comment by davidkyle** [comment]
> Another issue is that the Docker image size for the build image has gone up from 1.75GB to 23.9GB:

Removing the PyTorch repo and build artefacts has taken this size down to 5.4GB.

Removing python 3 will save another 2GB but I can't see away of doing this without putting all the instructions in a huge `RUN` statement. I will seek advice on the base way to do handle this.

**Comment by davidkyle** [comment]
> CMake has picked Acclerate for the BLAS library in that case. I will look into making that option explicit

On macOS the logic for picking a BLAS library is:

- If `BLAS` is set use the specified library and fail if it cannot be found
- Else default to `MKL`
- If `MKL` cannot be found use `Acclerate`
- Set `BLAS=vecLib` to explicitly choose the `Acclerate` framework

We should prefer Apple's libraries as opposed to a 3rd party ones and this may become more important with the `M1` build. Unfortunately there is a problem with setting `BLAS=vecLib` on recent versions of macOS as `veclib.h` will not be found. 

For the build servers this is not a problem because they do not have `MKL` installed so will naturally pick `Acclerate` if `BLAS` is unset. Developer machines will link `MKL` if it is installed.

**Comment by davidkyle** [comment]
> Removing python 3 will save another 2GB but I can't see away of doing this without putting all the instructions in a huge RUN statement. I will seek advice on the base way to do handle this.

The recommended method for removing intermediate artefacts is to use a multi-state docker file. The build framework is installed in a temporary image which is used to build the required libraries. Those copied to a fresh image which does not have any of the legacy build artefacts.
